Doc: Bluetooth: Mesh: Add RPR docs

Andrewpini opened the pull request:
Adds documentation for the Remote provisioning server and client
models.

ludvigsj commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,25 @@
+mesh specification. It is enabled with the
+:kconfig:option:`CONFIG_BT_MESH_RPR_CLI` option.
+
+The Remote Provisioning Client model is introduced in the Bluetooth Mesh Profile
```
```suggestion
The Remote Provisioning Client model is introduced in the Bluetooth Mesh Protocol
```
... same in srv

ludvigsj commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,25 @@
+:kconfig:option:`CONFIG_BT_MESH_RPR_CLI` option.
+
+The Remote Provisioning Client model is introduced in the Bluetooth Mesh Profile
+Specification version 1.1, and provides functionality to remotley provision devices into a mesh
```
```suggestion
Specification version 1.1, and provides functionality to remotely provision devices into a mesh
```
Same in srv

ludvigsj commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,25 @@
+Specification version 1.1, and provides functionality to remotley provision devices into a mesh
+network by interacting with mesh nodes that supports the :ref:`bluetooth_mesh_models_rpr_srv` model.
+
+The Remote Provisioning Client model communicates with an Remote Provisioning Server model
```
```suggestion
The Remote Provisioning Client model communicates with a Remote Provisioning Server model
```

ludvigsj commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,25 @@
+network by interacting with mesh nodes that supports the :ref:`bluetooth_mesh_models_rpr_srv` model.
+
+The Remote Provisioning Client model communicates with an Remote Provisioning Server model
+using the device key of the target node.
```
It might be confusing what "target node" is. The RPR server or the provisionee?

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/rpr_srv.rst:
```diff
@@ -0,0 +1,26 @@
+:ref:`bluetooth_mesh_models_rpr_cli` to control it. The Remote Provisioning Server
+model only accepts messages encrypted with the node's device key.
+
+If present, the Remote Provisioning Server model must be instantiated on the primary
+element. It may also be supported by any secondary element.
```
If I read the code correctly we only support single RPR Server instance which can only be instantiated on the primary element: https://github.com/zephyrproject-rtos/zephyr/blob/3e7205876cb5145d6630a7d1bcaf09945009f9f3/subsys/bluetooth/mesh/rpr_srv.c#L1306-L1310. So the second sentence can be removed.

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,25 @@
+The Remote Provisioning Client model communicates with a Remote Provisioning Server model
+using the device key of the node containing the target Remote Provisioning Server model instance.
+
+If present, the Remote Provisioning Client model must be instantiated on the primary
+element. It may also be supported by any secondary element.
```
We don't really support this as per my understanding. There is no check in `rpr_cli_init` that another RPR client instance was instantiate already, but even if we could instantiated several RPR clients, we can't provision multiple devices in parallel (we keep only one provisioning context). So I'd suggest to remove the second sentence and add check to `rpr_cli_init` in `rpr_cli.c` that element index is zero (like in server).

PavelVPV reviewed (changes requested):
Could we add a bit more information about what features these models provide (scanning, provisioning, re-provisioning) and how they are used (what API needs to be called and what happens after)?

Andrewpini commented:
> Could we add a bit more information about what features these models provide (scanning, provisioning, re-provisioning) and how they are used (what API needs to be called and what happens after)?

I'll take a look at it 👍

alxelax commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,25 @@
+
+The Remote Provisioning Client model is introduced in the Bluetooth Mesh Protocol
+Specification version 1.1, and provides functionality to remotely provision devices into a mesh
+network by interacting with mesh nodes that supports the :ref:`bluetooth_mesh_models_rpr_srv` model.
```
It adds more features like NPPI(or how it is called in the last spec version?) interface to change composition data, address and device key. Probably it is worth to add a bit more about these abilities since they are quite important. 
Also, it adds extended scanning (that includes active scanning) but I'm not sure if it is worth to go so deep.

alxelax reviewed (changes requested):
I guess it requires a bit more details about features that RPR adds. It is much more complex than just remote provisioning.

Andrewpini commented:
Food for thought: When reading through the spec the NPPI section states for all the 3 different procedures that one of the main benefits of the procedure is that it can change devkey, root_addr and comp_data **wihout** reprovisioning the device. However the API consequently uses the word `reprovision`. Is this just a simplification to let the user easily understand what the heck this function does? As far as the spec goes it is not strictly reprovisioning.

PavelVPV commented:
> Food for thought: When reading through the spec the NPPI section states for all the 3 different procedures that one of the main benefits of the procedure is that it can change devkey, root_addr and comp_data **wihout** reprovisioning the device.

I guess by "reprovisioning" the spec means unprovisioning via Node Reset message and provisioning the unprovisioned device again and NPPI is added specifically to avoid this.

> However the API consequently uses the word `reprovision`. Is this just a simplification to let the user easily understand what the heck this function does? As far as the spec goes it is not strictly reprovisioning. 

I think so. The API merges 3 procedures into 1 function and uses generic name for all 3 of them.

If this creates confusion from your point of view, feel free to rename. For example, "reprovision" could be renamed to "refresh".

Andrewpini commented:
> > Food for thought: When reading through the spec the NPPI section states for all the 3 different procedures that one of the main benefits of the procedure is that it can change devkey, root_addr and comp_data **wihout** reprovisioning the device.
> 
> I guess by "reprovisioning" the spec means unprovisioning via Node Reset message and provisioning the unprovisioned device again and NPPI is added specifically to avoid this.
> 
> > However the API consequently uses the word `reprovision`. Is this just a simplification to let the user easily understand what the heck this function does? As far as the spec goes it is not strictly reprovisioning.
> 
> I think so. The API merges 3 procedures into 1 function and uses generic name for all 3 of them.
> 
> If this creates confusion from your point of view, feel free to rename. For example, "reprovision" could be renamed to "refresh".

Im fine having it like it is, but I think I should leave parts of the spec definition out of the docs since it will be quite confusing. As long as it is not crucial for the user to know about the nuance difference between regular reprovison and the RPR reprovison procedure I think that will be fine.

alxelax commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,129 @@
+      uint8_t timeout = 10;
+      uint8_t max_devs = 3;
+
+      (void)bt_mesh_rpr_scan_start(&rpr_cli, &srv, uuid, timeout, max_devs, &status);
```
```suggestion
      bt_mesh_rpr_scan_start(&rpr_cli, &srv, uuid, timeout, max_devs, &status);
```
I guess the void conversion is not required for pseudo code.

alxelax commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,129 @@
+      uint16_t addr = 0x0006;
+      uint16_t net_idx = 0;
+
+      (void)bt_mesh_provision_remote(&rpr_cli, &srv, uuid, net_idx, addr);
```
```suggestion
      bt_mesh_provision_remote(&rpr_cli, &srv, uuid, net_idx, addr);
```

alxelax commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,129 @@
+      bool composition_changed = false;
+      uint16_t new_addr = 0x0009;
+
+      (void)bt_mesh_reprovision_remote(&rpr_cli, &srv, new_addr, composition_changed);
```
```suggestion
      bt_mesh_reprovision_remote(&rpr_cli, &srv, new_addr, composition_changed);
```

alxelax reviewed (commented):
LGTM just 3 minor technical comments

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,129 @@
+
+      bt_mesh_reprovision_remote(&rpr_cli, &srv, new_addr, composition_changed);
+
+The above example shows pseudo code for triggering a Node Address Refresh procedure on the target RPR server node.
```
```suggestion
The above example shows pseudo code for triggering a Node Address Refresh procedure on the target node.
```

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/rpr_srv.rst:
```diff
@@ -0,0 +1,26 @@
+
+If present, the Remote Provisioning Server model must be instantiated on the primary
+element.
+
```
Please add that when device key, node address or comp data are refreshed, `struct bt_mesh_prov.reprovisioned` callback is called.

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,129 @@
+The above example shows pseudo code for triggering a Node Address Refresh procedure on the target RPR server node.
+The specific procedure is not chosen directly, but rather through the other parameters that are inputted.
+In the example we can see that the current unicast address of the target is 0x0006, while the new address is
+set to 0x0009. If the two addresses had been the same, and the composition_changed flag set to true, this code
```
```suggestion
set to 0x0009. If the two addresses had been the same, and the ``composition_changed`` flag set to true, this code
```

PavelVPV commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,129 @@
+In the example we can see that the current unicast address of the target is 0x0006, while the new address is
+set to 0x0009. If the two addresses had been the same, and the composition_changed flag set to true, this code
+would instead have triggered a Node Composition Refresh procedure. If the two addresses had been the same, and
+the composition_changed flag set to false, this code would have triggered a Device Key Refresh procedure.
```
```suggestion
the ``composition_changed`` flag set to false, this code would have triggered a Device Key Refresh procedure.
```

PavelVPV reviewed (changes requested):
No more comments from my side. Will approve after @mia-ko review.

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+mesh specification. It is enabled with the
+:kconfig:option:`CONFIG_BT_MESH_RPR_CLI` option.
+
+The Remote Provisioning Client model is introduced in the Bluetooth Mesh Protocol
+Specification version 1.1, and provides functionality to remotely provision devices into a mesh
+network, as well as performing Node Provisioning Protocol Interface procedures, by interacting with mesh nodes that supports the :ref:`bluetooth_mesh_models_rpr_srv` model.
```
```suggestion
The Remote Provisioning Client model is introduced in the Bluetooth Mesh Protocol
Specification version 1.1.
This model provides functionality to remotely provision devices into a mesh network, and perform Node Provisioning Protocol Interface procedures by interacting with mesh nodes that support the :ref:`bluetooth_mesh_models_rpr_srv` model.
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+
+      bt_mesh_rpr_scan_start(&rpr_cli, &srv, uuid, timeout, max_devs, &status);
+
+The above example shows pseudo code for starting a scan procedure on the target RPR server node. This
```
```suggestion
The above example shows pseudo code for starting a scan procedure on the target Remote Provisioning Server node. This
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+      bt_mesh_rpr_scan_start(&rpr_cli, &srv, uuid, timeout, max_devs, &status);
+
+The above example shows pseudo code for starting a scan procedure on the target RPR server node. This
+procedure will start a ten second multiple devices scanning where the generated scan report will contain
```
```suggestion
procedure will start a ten-second, multiple-device scanning where the generated scan report will contain
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+
+The above example shows pseudo code for starting a scan procedure on the target RPR server node. This
+procedure will start a ten second multiple devices scanning where the generated scan report will contain
+a maximum of three unprovisioned device. If the UUID argument was specified, the same procedure would
```
```suggestion
a maximum of three unprovisioned devices. If the UUID argument was specified, the same procedure would
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+The above example shows pseudo code for starting a scan procedure on the target RPR server node. This
+procedure will start a ten second multiple devices scanning where the generated scan report will contain
+a maximum of three unprovisioned device. If the UUID argument was specified, the same procedure would
+only scan for the device with the corresponding UUID. After the procedure completes the
+server will send the scan report that will be handled in the clients :c:member:`bt_mesh_rpr_cli.scan_report` callback.
```
```suggestion
only scan for the device with the corresponding UUID. After the procedure completes, the
server sends the scan report that will be handled in the client's :c:member:`bt_mesh_rpr_cli.scan_report` callback.
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+
+      bt_mesh_provision_remote(&rpr_cli, &srv, uuid, net_idx, addr);
+
+The above example shows pseudo code for remotely provisioning a device through a RPR server node. This
```
```suggestion
The above example shows pseudo code for remotely provisioning a device through a Remote Provisioning Server node. This
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+procedure will attempt to provision the device with the corresponding UUID, and assign the address 0x0006
+to its primary element using the network key located at index zero.
+
+Please note that during remote provisioning the same :c:struct:`bt_mesh_prov` callbacks are triggered as for ordinary
+provisioning. Please refer to section :ref:`bluetooth_mesh_provisioning` for further details.
```
```suggestion
Note that during the remote provisioning, the same :c:struct:`bt_mesh_prov` callbacks are triggered as for ordinary
provisioning. See section :ref:`bluetooth_mesh_provisioning` for further details.
```

mia-ko replied:
Also, should this be a note?

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+Please note that during remote provisioning the same :c:struct:`bt_mesh_prov` callbacks are triggered as for ordinary
+provisioning. Please refer to section :ref:`bluetooth_mesh_provisioning` for further details.
+
+Re-Provisioning
```
I am not sure if Zephyr doc uses title case for headings, there are too many inconsistencies to actually tell. So we stick to what we know from NCS (sentence case in headings).
```suggestion
Re-provisioning
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+***************
+
+In addition to scanning and provisioning functionality, the Remote Provisioning Client also provides means to
+reconfigure node addresses, device keys and composition data on devices that support the
```
```suggestion
reconfigure node addresses, device keys and Composition Data on devices that support the
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+:ref:`bluetooth_mesh_models_rpr_srv` model. This is provided through the Node Provisioning Protocol Interface
+(NPPI) which supports the following three procedures:
+
+- Device Key Refresh procedure: Used to change the device key of the target node without a need to reconfigure the node.
+- Node Address Refresh procedure: Used to change the node’s device key and unicast address.
+- Node Composition Refresh procedure: Used to change the device key of the node and to add or delete models or features
+  of the node.
```
```suggestion
* Device Key Refresh procedure: Used to change the device key of the Target node without a need to reconfigure the node.
* Node Address Refresh procedure: Used to change the node’s device key and unicast address.
* Node Composition Refresh procedure: Used to change the device key of the node, and to add or delete models or features of the node.
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+
+      bt_mesh_reprovision_remote(&rpr_cli, &srv, new_addr, composition_changed);
+
+The above example shows pseudo code for triggering a Node Address Refresh procedure on the target node.
```
```suggestion
The above example shows pseudo code for triggering a Node Address Refresh procedure on the Target node.
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+
+The above example shows pseudo code for triggering a Node Address Refresh procedure on the target node.
+The specific procedure is not chosen directly, but rather through the other parameters that are inputted.
+In the example we can see that the current unicast address of the target is 0x0006, while the new address is
```
```suggestion
In the example we can see that the current unicast address of the Target is 0x0006, while the new address is
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_cli.rst:
```diff
@@ -0,0 +1,132 @@
+The above example shows pseudo code for triggering a Node Address Refresh procedure on the target node.
+The specific procedure is not chosen directly, but rather through the other parameters that are inputted.
+In the example we can see that the current unicast address of the target is 0x0006, while the new address is
+set to 0x0009. If the two addresses had been the same, and the ``composition_changed`` flag set to true, this code
+would instead have triggered a Node Composition Refresh procedure. If the two addresses had been the same, and
+the ``composition_changed`` flag set to false, this code would have triggered a Device Key Refresh procedure.
```
```suggestion
set to 0x0009. If the two addresses were the same, and the ``composition_changed`` flag was set to true, this code
would instead trigger a Node Composition Refresh procedure. If the two addresses were the same, and
the ``composition_changed`` flag was set to false, this code would trigger a Device Key Refresh procedure.
```

mia-ko commented on doc/connectivity/bluetooth/api/mesh/rpr_srv.rst:
```diff
@@ -0,0 +1,30 @@
+If present, the Remote Provisioning Server model must be instantiated on the primary
+element.
+
+Please note that after refreshing the device key, node address or composition data through a Node Provisioning Protocol
+Interface (NPPI) procedure, the :c:member:`bt_mesh_prov.reprovisioned` callback is triggered. Please refer to section
+:ref:`bluetooth_mesh_models_rpr_cli` for further details.
```
```suggestion
Note that after refreshing the device key, node address or Composition Data through a Node Provisioning Protocol
Interface (NPPI) procedure, the :c:member:`bt_mesh_prov.reprovisioned` callback is triggered. See section
:ref:`bluetooth_mesh_models_rpr_cli` for further details.
```

Andrewpini committed:
Bluetooth: Mesh: Change rpr cli @defgroup name

Change the documentation @defgroup tag for RPR client,
as it does not align with other parts of the RPR implementation.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Bluetooth: Mesh: RPR cli @ mod elem 0 check

Add check to the Remote Provisioning Client model init call
that verifies that the model is located at the root element.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
Doc: Bluetooth: Mesh: Add RPR docs

Adds documentation for the Remote provisioning server and client
models.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>
